hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/filter/ReclaimableKeyFilter.java

```diff
@@ -0,0 +1,275 @@
+    if (deletedKeyInfo.getObjectID() == OBJECT_ID_RECLAIM_BLOCKS) {
+      return true;
+    }
+
+    // Construct keyTable or fileTable DB key depending on the bucket type
+    if (bucketInfo.getBucketLayout().isFileSystemOptimized()) {
+      dbKey = ozoneManager.getMetadataManager().getOzonePathKey(
+          volumeId,
+          bucketInfo.getObjectID(),
+          deletedKeyInfo.getParentObjectID(),
+          deletedKeyInfo.getFileName());
+    } else {
+      dbKey = ozoneManager.getMetadataManager().getOzoneKey(
+          deletedKeyInfo.getVolumeName(),
+          deletedKeyInfo.getBucketName(),
+          deletedKeyInfo.getKeyName());
+    }
+
+    /*
+     snapshotRenamedTable:
+     1) /volumeName/bucketName/objectID ->
+                 /volumeId/bucketId/parentId/fileName (FSO)
+     2) /volumeName/bucketName/objectID ->
+                /volumeName/bucketName/keyName (non-FSO)
+    */
+    String dbRenameKey = ozoneManager.getMetadataManager().getRenameKey(
+        deletedKeyInfo.getVolumeName(), deletedKeyInfo.getBucketName(),
+        deletedKeyInfo.getObjectID());
+
+    // Condition: key should not exist in snapshotRenamedTable
+    // of the current snapshot and keyTable of the previous snapshot.
+    // Check key exists in renamedTable of the Snapshot
+    String renamedKey = renamedTable.getIfExist(dbRenameKey);
+
+    if (renamedKey != null && renamedKeyBuilder != null) {
+      renamedKeyBuilder.setKey(dbRenameKey).setValue(renamedKey);
+    }
+    // previousKeyTable is fileTable if the bucket is FSO,
+    // otherwise it is the keyTable.
+    OmKeyInfo prevKeyInfo = renamedKey != null ? previousKeyTable
+        .get(renamedKey) : previousKeyTable.get(dbKey);
+
+    if (prevKeyInfo == null ||
+        prevKeyInfo.getObjectID() != deletedKeyInfo.getObjectID()) {
+      return true;
+    }
+
+    // For key overwrite the objectID will remain the same, In this
+    // case we need to check if OmKeyLocationInfo is also same.
+    return !isBlockLocationInfoSame(prevKeyInfo, deletedKeyInfo);
+  }
+
+  /**
+   * To calculate Exclusive Size for current snapshot, Check
+   * the next snapshot deletedTable if the deleted key is
+   * referenced in current snapshot and not referenced in the
+   * previous snapshot then that key is exclusive to the current
+   * snapshot. Here since we are only iterating through
+   * deletedTable we can check the previous and previous to
+   * previous snapshot to achieve the same.
+   * previousSnapshot - Snapshot for which exclusive size is
+   *                    getting calculating.
+   * currSnapshot - Snapshot's deletedTable is used to calculate
+   *                previousSnapshot snapshot's exclusive size.
+   * previousToPrevSnapshot - Snapshot which is used to check
+   *                 if key is exclusive to previousSnapshot.
+   */
+  @SuppressWarnings("checkstyle:ParameterNumber")
+  public void calculateExclusiveSize(
```

Comment:
Should it be part of the reclaimable filter?

Reply:
yeah we would want to calculate the exclusive size as part of the keyDeletingService run itself. Don't need a separate thread for this.